Changing envelope definition to  support two fixed modes

kevinlewi opened the pull request:
See #99.

This defines two modes, `base` and `customIdentifier` for the envelope. The credentials are incorporated in the `SecretCredentials` and `CleartextCredentials` structs,
depending on the mode set by the value of `EnvelopeMode`:

```
enum {
  base(1),
  customIdentifier(2),
  (255)
} EnvelopeMode;
```

The `base` mode defines `SecretCredentials` and `CleartextCredentials` as follows:

```
struct {
  opaque skU<1..2^16-1>;
} SecretCredentials;

struct {
  opaque pkS<1..2^16-1>;
} CleartextCredentials;
```

The `customIdentifier` mode defines `SecretCredentials` and `CleartextCredentials` as follows:

```
struct {
  opaque skU<1..2^16-1>;
} SecretCredentials;

struct {
  opaque pkS<1..2^16-1>;
  opaque idU<0..2^16-1>;
  opaque idS<0..2^16-1>;
} CleartextCredentials;
```

These credentials are embedded into the following `Envelope` structure with
encryption and authentication.

```
struct {
  InnerEnvelopeMode mode;
  opaque nonce[32];
  opaque ct<1..2^16-1>;
} InnerEnvelope;

struct {
  InnerEnvelope contents;
  opaque auth_tag[Nh];
} Envelope;
```

Also updated the `FinalizeRequest` and `RecoverCredentials` procedures accordingly. Also updated the fixed string to read `OPAQUE01` instead of `OPAQUE00`.

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -465,13 +459,14 @@ nonce
 auth_tag
-: Authentication tag protecting the contents of the envelope.
+: Authentication tag protecting the contents of the envelope,
+incorporating `InnerEnvelope` and `CleartextCredentials`.
```
```suggestion
covering `InnerEnvelope` and `CleartextCredentials`.
```

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -1038,8 +1012,8 @@ info = "HMQV keys" || I2OSP(len(nonceU), 2) || nonceU
-`CredentialExtension` for the envelope; however, if no such extension were supplied,
-then these values are defaulted to pkU and pkS instead.
+credential for the envelope; however, if no such credential was supplied,
+then these values are (indepedently) defaulted to pkU and pkS instead.
```
```suggestion
then these values (independently) default to pkU and pkS instead.
```

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -465,13 +459,14 @@ nonce
+incorporating `InnerEnvelope` and `CleartextCredentials`.
+
+The full procedure for constructing `Envelope` and `InnerEnvelope` from
+`SecretCredentials` and `CleartextCredentials` is described in {{finalize-request}}.
```
```suggestion
`SecretCredentials` and `CleartextCredentials` is described in {{finalize-request}}.

The `EnvelopeMode` value is specified as part of a configuration. 
```

chris-wood reviewed (commented):
I like this :) I have a couple questions that I hope we can resolve moving forward:

- The mode is currently implicit in registration and login. I assume the expectation is that configurations specify the mode so that it's known to both peers. We should probably say that somewhere. I left a suggestion to include this note, but we probably ought to expand upon it in the configuration section.
- The `RegistrationResponse` definition still includes a list of secret and cleartext credential types. This can be removed given that these are dependent on the mode. 
- Credential construction is somewhat vague in the current description, especially since idU and idS are not passed as parameters to the main registration and login routines. As an implementor, is the expectation that idS comes from the client's wrapper protocol, or is it sent from server to client as part of the registration response? (My understanding is the latter.) Since there are only two modes, we might, for example, define two mode-specific `FinalizeRequest` variants, each of which takes the appropriate parameters. (The base mode variant includes what is currently there now, whereas the custom-mode variant might also include idU (and idS depending on the answer to the previous question).)

Obviously, we can work on these things incrementally and merge in pieces, but I wanted to flag these questions to start.

kevinlewi commented:
* I added a note in the configurations section: "The EnvelopeMode value is defined in {{data-types}} and must also be agreed upon between the client and server."
* Thanks, good catch. Removed!
* The expectation is that idS should come from the wrapper protocol. I think it is unnecessary to include idS in `RegistrationResponse`, regardless of the mode. Therefore, I think it is unnecessary to change anything here regarding this point. What do you think?

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -1174,6 +1148,9 @@ The OPAQUE AKE protocols are those which are specified in {{instantiations}}.
 Future specifications (such as {{I-D.sullivan-tls-opaque}}) MAY introduce other
 AKE instantiations.
 
+The EnvelopeMode value is defined in {{data-types}} and must also be agreed upon
```
This implies (to me) some sort of negotiation. Since the EnvelopeMode is already in the tuple above, I'd drop that bit:

```suggestion
The EnvelopeMode value is defined in {{data-types}}.
```

chris-wood commented:
> The expectation is that idS should come from the wrapper protocol. I think it is unnecessary to include idS in RegistrationResponse, regardless of the mode. Therefore, I think it is unnecessary to change anything here regarding this point. What do you think?

I think that's fine, but I would make things more explicit. If the wrapper protocol provides them, then it might as well also supply them to the relevant procedures for constructing the envelope. Then we could replace this:

"Create cleartext_credentials with credentials matching those needed to construct the `CleartextCredentials` structure"

with something like:

"Create `CleartextCredentials` structure cleartext_credentials where cleartext_credentials.idU = idU, cleartext_credentials.idS = idS, cleartext_credentials.pkS = pkS."

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -624,45 +612,31 @@ Output:
+13. Create InnerEnvelope contents with (nonce, ct)
+14. t = HMAC(auth_key, concat(contents, auth_data))
 15. Create Envelope envU with (contents, t)
 16. Create RegistrationUpload upload with envelope value (envU, pkU)
```
Unrelated: should we lift `pkU` to the wrapper protocol, too?

kevinlewi replied:
As discussed offline, let's address this API change separately.

chris-wood commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -598,11 +586,11 @@ Steps:
+#### FinalizeRequest {#finalize-request}
 
 ~~~
 FinalizeRequest(pwdU, skU, blind, request, response)
```
Also unrelated, but we can remove `request` here, as it's not used.

kevinlewi replied:
It is used in line 468:

```
The full procedure for constructing `Envelope` and `InnerEnvelope` from
`SecretCredentials` and `CleartextCredentials` is described in {{finalize-request}}.
```

chris-wood replied:
I'll address this in a separate PR, since it seems related to the module boundary we discussed.

chris-wood commented:
Following up on the comment above, we might simplify things by delegating the "credential construction" logic to the wrapper protocol, which (a) knows the mode and (b) has the relevant information (skU, idU, pkS, idS, etc). That would make the interface to `FinalizeRequest` something like so:

```
FinalizeRequest(pwdU, creds, blind, response)
```

Where `creds` contains the `secret_credentials` and `cleartext_credentials`, already in wire format form, to use when creating the envelope.

bytemare commented:
Since idu is used on both sides for the AKE info HKDF parameter, shouldn't it always be conveyed to the client?

bytemare commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -624,45 +612,31 @@ Output:
+   construct the `SecretCredentials` structure
+5. Create cleartext_credentials with credentials matching those needed to
+   construct the `CleartextCredentials` structure
+6. pt = Serialize(secret_credentials)
```
If SecretCredentials is always equal to the client secret, wouldn't it be more straightforward to put `pt = Serialize(secret)` ?

kevinlewi replied:
I think it may be beneficial to keep this level of generality here in case we add support for other modes that set SecretCredentials to something else in the future. This will hopefully also coincide with the changes that will be made in #105 which will pass in a generic credentials parameter.

bytemare commented on draft-irtf-cfrg-opaque.md:
```diff
@@ -624,45 +612,31 @@ Output:
-13. Create InnerEnvelope contents with (nonce, ct, auth_data)
-14. t = HMAC(auth_key, contents)
+12. auth_data = Serialize(cleartext_credentials)
+13. Create InnerEnvelope contents with (nonce, ct)
```
If I'm not mistaken, the InnerEnvelope should also specify the InnerEnvelopeMode, doesn't it?

kevinlewi replied:
Good catch, thank you! Added to parameters.

kevinlewi commented:
Addressed comments, ready for a re-review
